### PR TITLE
throw more specific toolexit when git fails during upgrade

### LINKED ISSUE #56853: `flutter upgrade` gives wrong error with detached head: "Unable to upgrade Flutter: no origin repository configured"
## Steps to Reproduce

1. upgrade from v1.12.13+hotfix.9 to 1.17.0
2. use flutter downgrade to v1.12.13+hotfix.9
3. run on iOS

- I tried to delete the bin/cache directory and do flutter doctor, but the issue is still there
- and if I run flutter upgrade , it shows Unable to upgrade Flutter: no origin repository configured. Run 'git remote add origin https://github.com/flutter/flutter' in /Users/apple/Flutter/flutter,
I tried flutter channel stable , not works. please help me 

## Logs

```
Xcode build done.                                           29.4s
[VERBOSE-2:shell.cc(199)] Dart Error: Can't load Kernel binary: Invalid kernel binary format version.
[VERBOSE-2:engine.cc(198)] Could not prepare to run the isolate.
[VERBOSE-2:engine.cc(137)] Engine not prepare and launch isolate.
[VERBOSE-2:shell.cc(437)] Could not launch engine with configuration.
```

```
[✓] Flutter (Channel unknown, v1.12.13+hotfix.9, on Mac OS X 10.14.6 18G87, locale en-US)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
[✓] Android Studio (version 3.6)
[✓] VS Code (version 1.44.2)
[✓] Connected device (2 available)
```

### PR DESCRIPTION
**Pull request by christopherfujino**
## Description

If the user's git repo is in a detached HEAD state (which happens after using `flutter version <tag>` or `git checkout <tag>`) and they try to `flutter upgrade`, they will receive in an incorrect error message: `Unable to upgrade Flutter: no origin repository configured`.

This PR catches the process exception, and throws specific tool exits if either their head is detached or they don't have an upstream configured, else the original git error message is surfaced.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/56853.

## Tests

I added two new unit tests to validate the new tool exit messages.

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/upgrade.dart [hunk 1]
```diff
@@ -219,12 +219,21 @@ class UpgradeCommandRunner {
...
       revision = result.stdout.trim();
-    } on Exception {
-      throwToolExit(
-        'Unable to upgrade Flutter: no origin repository configured. '
-        "Run 'git remote add origin "
-        "https://github.com/flutter/flutter' in $workingDirectory",
-      );
+    } on Exception catch (e){
+      final String errorString = e.toString();
+      if (RegExp(r'fatal: HEAD does not point to a branch').hasMatch(errorString)) {
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/commands/upgrade.dart:224]: `errorString.contains`

File: packages/flutter_tools/lib/src/commands/upgrade.dart [hunk 2]
```diff
@@ -219,12 +219,21 @@ class UpgradeCommandRunner {
...
+        throwToolExit(
+          'You are not currently on a release branch. Use git to '
+          'check out an official branch\n(\'stable\', \'beta\', \'dev\', or \'master\') '
+          'and retry, for example \'git checkout stable\'.'
+        );
+      } else if (RegExp(r'fatal: no upstream configured for branch').hasMatch(errorString)) {
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/commands/upgrade.dart:230]: `contains`

File: packages/flutter_tools/lib/src/commands/upgrade.dart [hunk 3]
```diff
@@ -219,12 +219,21 @@ class UpgradeCommandRunner {
           workingDirectory: workingDirectory,
       );
       revision = result.stdout.trim();
-    } on Exception {
-      throwToolExit(
-        'Unable to upgrade Flutter: no origin repository configured. '
-        "Run 'git remote add origin "
-        "https://github.com/flutter/flutter' in $workingDirectory",
-      );
+    } on Exception catch (e){
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/commands/upgrade.dart:222]: Missing space between `(` and `{`

File: packages/flutter_tools/lib/src/commands/upgrade.dart [hunk 4]
```diff
@@ -219,12 +219,21 @@ class UpgradeCommandRunner {
...
+      final String errorString = e.toString();
+      if (RegExp(r'fatal: HEAD does not point to a branch').hasMatch(errorString)) {
+        throwToolExit(
+          'You are not currently on a release branch. Use git to '
+          'check out an official branch\n(\'stable\', \'beta\', \'dev\', or \'master\') '
+          'and retry, for example \'git checkout stable\'.'
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/commands/upgrade.dart:228]: Based on an audit of user messages, how about:
```
'You are not currently on a release branch. Use git to '
'check out an official branch (\'stable\', \'beta\', \'dev\', or \'master\') '
'and retry, for example:\n'
'  git checkout stable'
```

File: packages/flutter_tools/lib/src/commands/upgrade.dart (hunk 1 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/upgrade.dart:224]: done

File: packages/flutter_tools/lib/src/commands/upgrade.dart (hunk 2 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/upgrade.dart]: done

File: packages/flutter_tools/lib/src/commands/upgrade.dart (hunk 3 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/upgrade.dart]: done

File: packages/flutter_tools/lib/src/commands/upgrade.dart (hunk 4 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/upgrade.dart]: done

**Review by jmagman — APPROVED** [review]
LGTM

**Pull request merged**